File: generate-images.js
require('dotenv').config();
const fs = require('fs').promises;
const path = require('path');
const fetch = require('node-fetch');

const engineId = 'stable-diffusion-xl-1024-v1-0';
const apiHost = 'https://api.stability.ai';

// 곤충 데이터
const insects = [
    {
        name: "호랑나비",
        category: "butterflies",
        prompt: "Create exactly one single Korean Yellow Swallowtail (Papilio xuthus) in vintage scientific illustration style. Strict requirement: only one butterfly in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle yellow and black watercolor washes. The butterfly must be positioned in dorsal view with wings fully spread, showing detailed anatomical features: compound eyes, coiled proboscis, segmented antennae, detailed wing patterns with clear veination, thorax structure, and articulated legs. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "무당벌레",
        category: "beetles",
        prompt: "Create exactly one single Seven-spotted Ladybug (Coccinella septempunctata) in vintage scientific illustration style. Strict requirement: only one ladybug in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle red and black watercolor washes. The ladybug must be positioned in dorsal view with elytra clearly visible, showing detailed anatomical features: compound eyes, segmented antennae, precisely seven black spots on red wing covers, thorax markings, and articulated legs. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "장수풍뎅이",
        category: "beetles",
        prompt: "Create exactly one single Japanese Rhinoceros Beetle (Allomyrina dichotoma) in vintage scientific illustration style. Strict requirement: only one beetle in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle brown and black watercolor washes. The beetle must be positioned in side view showing the horn profile, showing detailed anatomical features: distinctive horn structure, compound eyes, segmented antennae, textured elytra, powerful legs with visible joints, and thorax detail. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "사마귀",
        category: "grasshoppers",
        prompt: "Create exactly one single Praying Mantis (Tenodera sinensis) in vintage scientific illustration style. Strict requirement: only one mantis in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle green and brown watercolor washes. The mantis must be positioned in side view with raptorial forelegs visible, showing detailed anatomical features: triangular head, compound eyes, distinctive prayer-like forelegs, wing detail, segmented abdomen, and thorax structure. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "매미",
        category: "bees_moths",
        prompt: "Create exactly one single Cicada (Cryptotympana dubia) in vintage scientific illustration style. Strict requirement: only one cicada in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle brown and black watercolor washes. The cicada must be positioned in dorsal view with wings spread, showing detailed anatomical features: large compound eyes, transparent wing membranes with clear venation, broad head, segmented body, and sturdy legs. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "잠자리",
        category: "dragonflies",
        prompt: "Create exactly one single Korean Dragonfly (Anax parthenope) in vintage scientific illustration style. Strict requirement: only one dragonfly in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle blue and brown watercolor washes. The dragonfly must be positioned in dorsal view with all four wings spread, showing detailed anatomical features: large compound eyes, intricate wing venation, long segmented abdomen, thorax structure, and detailed leg joints. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "꿀벌",
        category: "bees_moths",
        prompt: "Create exactly one single Asian Honeybee (Apis cerana) in vintage scientific illustration style. Strict requirement: only one bee in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle golden-brown and black watercolor washes. The bee must be positioned in side view with wings slightly spread, showing detailed anatomical features: compound eye, segmented antennae, articulated legs with pollen baskets, fuzzy thorax, striped abdomen, and transparent wing membranes with clear venation. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "귀뚜라미",
        category: "grasshoppers",
        prompt: "Create exactly one single Field Cricket (Gryllus bimaculatus) in vintage scientific illustration style. Strict requirement: only one cricket in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle brown and black watercolor washes. The cricket must be positioned in side view with long antennae extended, showing detailed anatomical features: prominent jumping legs, wing structure, segmented antennae, cerci, and detailed leg joints. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "방아깨비",
        category: "grasshoppers",
        prompt: "Create exactly one single Rice Grasshopper (Oxya chinensis) in vintage scientific illustration style. Strict requirement: only one grasshopper in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle green and brown watercolor washes. The grasshopper must be positioned in side view with jumping legs extended, showing detailed anatomical features: compound eyes, segmented antennae, powerful hind legs, wing structure, and segmented abdomen. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    },
    {
        name: "하늘소",
        category: "beetles",
        prompt: "Create exactly one single Asian Long-horned Beetle (Anoplophora glabripennis) in vintage scientific illustration style. Strict requirement: only one beetle in the entire image. Use precise fine black ink lines to define form and structure, then apply subtle black and white watercolor washes. The beetle must be positioned in side view with long antennae extended, showing detailed anatomical features: distinctive long antennae, compound eyes, textured elytra with spots, segmented legs, and thorax structure. Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. Pure white background with absolutely no decorative elements. Reference style: 19th century entomological illustration techniques.",
        negative_prompt: "multiple insects, groups, pairs, pattern, collection, collage, specimen plate, decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, repeating elements, mirror images, kaleidoscope effect"
    }
];

async function generateImage(insect) {
    try {
        console.log(`Generating image for ${insect.name}...`);

        const response = await fetch(
            `${apiHost}/v1/generation/${engineId}/text-to-image`,
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    Accept: 'application/json',
                    Authorization: `Bearer ${process.env.STABILITY_API_KEY}`,
                },
                body: JSON.stringify({
                    text_prompts: [
                        {
                            text: insect.prompt,
                            weight: 1
                        },
                        ...(insect.negative_prompt ? [{
                            text: insect.negative_prompt,
                            weight: -1
                        }] : [])
                    ],
                    cfg_scale: parseInt(process.env.IMAGE_CFG_SCALE),
                    steps: parseInt(process.env.IMAGE_STEPS),
                    width: 1024,
                    height: 1024,
                    samples: 1,
                }),
            }
        );

        if (!response.ok) {
            throw new Error(`Non-200 response: ${await response.text()}`);
        }

        const responseJSON = await response.json();
        
        // 이미지 데이터 (base64)
        const image = responseJSON.artifacts[0];
        
        // 이미지 저장 경로 설정
        const imagePath = path.join(__dirname, 'images', insect.category, `${insect.name}.png`);
        
        // 이미지 저장
        await fs.writeFile(imagePath, Buffer.from(image.base64, 'base64'));
        console.log(`Image saved for ${insect.name} at ${imagePath}`);
        
        // 생성 사이에 잠시 대기
        await new Promise(resolve => setTimeout(resolve, 2000));

    } catch (error) {
        console.error(`Error generating image for ${insect.name}:`, error);
    }
}

// 이미지 생성 시작
const insectsToGenerate = ["꿀벌"];
const selectedInsects = insects.filter(insect => insectsToGenerate.includes(insect.name));

generateAllImages().then(() => {
    console.log('Selected images generated successfully!');
}).catch(error => {
    console.error('Error generating images:', error);
});

async function generateAllImages() {
    for (const insect of selectedInsects) {
        await generateImage(insect);
    }
} 

File: generate_images.py
import os
import requests
import base64
from PIL import Image
import io
import traceback

# Stable Diffusion API configuration
API_HOST = 'https://api.stability.ai'
API_KEY = 'sk-YMj12sqndEvoYLCJYBnq7bkZvyxBZfyVebALFM2d5MdTKKp6'

# Ensure the output directory exists
os.makedirs('images/not-insects', exist_ok=True)
os.makedirs('images/insects', exist_ok=True)

def generate_image(name, details, is_insect=False):
    try:
        prompt = f"""Create exactly one single {details['name']} in vintage scientific illustration style. 
        Strict requirement: only one specimen in the entire image. 
        Use precise fine black ink lines to define form and structure, then apply subtle {details['color']} watercolor washes. 
        The specimen must be positioned in {details['view']}, showing detailed anatomical features: {details['features']}. 
        Match the traditional natural history illustration style with meticulous line work and controlled watercolor technique. 
        Pure white background with absolutely no decorative elements. 
        Reference style: 19th century entomological illustration techniques."""

        negative_prompt = """multiple specimens, groups, pairs, pattern, collection, collage, specimen plate, 
        decorative border, text, labels, numbers, botanical elements, leaves, flowers, creative interpretation, 
        modern style, digital effects, filters, heavy shading, dark colors, artistic freedom, symmetrical arrangement, 
        repeating elements, mirror images, kaleidoscope effect"""

        print(f"Sending request for {name}...")
        
        response = requests.post(
            f"{API_HOST}/v1/generation/stable-diffusion-xl-1024-v1-0/text-to-image",
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
                "Authorization": f"Bearer {API_KEY}"
            },
            json={
                "text_prompts": [
                    {
                        "text": prompt,
                        "weight": 1
                    },
                    {
                        "text": negative_prompt,
                        "weight": -1
                    }
                ],
                "cfg_scale": 7,
                "height": 1024,
                "width": 1024,
                "samples": 1,
                "steps": 50,
            },
        )

        print(f"Response status code: {response.status_code}")
        print(f"Response content: {response.text[:500]}...")

        if response.status_code != 200:
            raise Exception(f"Non-200 response: {response.text}")

        data = response.json()
        
        # Save the image
        for i, image in enumerate(data["artifacts"]):
            image_data = base64.b64decode(image["base64"])
            folder = "insects" if is_insect else "not-insects"
            image_path = f"images/{folder}/{name}.png"
            
            # Open and save the image using PIL
            img = Image.open(io.BytesIO(image_data))
            img.save(image_path, format="PNG", quality=95)
            print(f"Generated and saved: {image_path}")

    except Exception as e:
        print(f"Error generating {name}:")
        print(traceback.format_exc())
        raise

# List of non-insect animals
not_insects = {
    '거미': {
        'name': 'European garden spider (Araneus diadematus)',
        'view': 'dorsal view with legs spread naturally',
        'features': 'eight legs with visible joints, cephalothorax and abdomen clearly separated, detailed eye pattern with 8 eyes, spinnerets visible, fine hairs on legs',
        'color': 'brown'
    },
    '지네': {
        'name': 'house centipede (Scutigera coleoptrata)',
        'view': 'lateral view with body slightly curved',
        'features': 'long antennae, 15 pairs of legs with characteristic length gradation, segmented body structure, modified first leg pair as forcipules',
        'color': 'reddish-brown'
    },
    '문어': {
        'name': 'common octopus (Octopus vulgaris)',
        'view': 'frontal view with tentacles spread',
        'features': 'eight arms with visible suction cups, large eyes, textured skin with chromatophores, funnel visible, mantle structure clear',
        'color': 'reddish'
    },
    '전갈': {
        'name': 'emperor scorpion (Pandinus imperator)',
        'view': 'dorsal view with tail curved',
        'features': 'large pincers, eight legs, segmented tail with visible stinger, detailed carapace texture, book lungs visible',
        'color': 'black'
    },
    '달팽이': {
        'name': 'garden snail (Helix pomatia)',
        'view': 'lateral view while crawling',
        'features': 'spiral shell with growth lines, extended foot, tentacles with eyes, visible mantle and pneumostome',
        'color': 'brown'
    },
    '지렁이': {
        'name': 'common earthworm (Lumbricus terrestris)',
        'view': 'lateral view with body gently curved',
        'features': 'clearly visible segments, clitellum, setae, anterior and posterior ends distinct, surface texture with subtle rings',
        'color': 'pink'
    },
    '게': {
        'name': 'shore crab (Carcinus maenas)',
        'view': 'dorsal view with legs spread',
        'features': 'carapace texture, ten legs including claws, compound eyes on stalks, segmented legs, detailed shell patterns',
        'color': 'orange-red'
    },
    '불가사리': {
        'name': 'common starfish (Asterias rubens)',
        'view': 'dorsal view with arms spread evenly',
        'features': 'five arms with tube feet visible, central disc, madreporite, spines and ossicles detailed, textured surface',
        'color': 'orange'
    },
    '해파리': {
        'name': 'moon jellyfish (Aurelia aurita)',
        'view': 'lateral view while swimming',
        'features': 'bell with detailed margin, four oral arms, radial canals, gonads visible through translucent bell, tentacles around bell margin',
        'color': 'translucent blue'
    },
    '쥐며느리': {
        'name': 'common woodlouse (Armadillidium vulgare)',
        'view': 'dorsal view',
        'features': 'segmented body armor, seven pairs of legs, antennae, compound eyes, detailed plate structure, uropods visible',
        'color': 'grey'
    }
}

# List of insects
insects = {
    '호랑나비': {
        'name': 'Asian swallowtail butterfly (Papilio xuthus)',
        'view': 'dorsal view with wings fully spread',
        'features': 'large yellow wings with black tiger-like stripes, distinctive swallowtail extensions, detailed wing veins, slender body, long antennae, compound eyes',
        'color': 'yellow and black'
    },
    '무당벌레': {
        'name': 'Seven-spotted ladybug (Coccinella septempunctata)',
        'view': 'dorsal view showing wing covers',
        'features': 'round body, distinctive spots on wing covers, small head with antennae, six legs visible, dome-shaped body',
        'color': 'red with black spots'
    },
    '장수풍뎅이': {
        'name': 'Japanese rhinoceros beetle (Allomyrina dichotoma)',
        'view': 'lateral view showing horn',
        'features': 'large horn on head, strong legs with spines, detailed exoskeleton texture, wing covers, antennae',
        'color': 'brown to black'
    },
    '사마귀': {
        'name': 'Praying mantis (Tenodera sinensis)',
        'view': 'lateral view in prayer-like pose',
        'features': 'large compound eyes, distinctive raptorial forelegs, long body, detailed wing texture, segmented abdomen',
        'color': 'green'
    },
    '매미': {
        'name': 'Cicada (Cryptotympana facialis)',
        'view': 'dorsal view with wings spread',
        'features': 'transparent wings with detailed veins, large compound eyes, broad head, sturdy body, distinctive wing pattern',
        'color': 'brown with transparent wings'
    },
    '잠자리': {
        'name': 'Common skimmer dragonfly (Orthetrum albistylum)',
        'view': 'dorsal view with wings spread horizontally',
        'features': 'four transparent wings with intricate veins, long slender body, large compound eyes, detailed thorax structure',
        'color': 'blue with clear wings'
    },
    '꿀벌': {
        'name': 'Western honey bee (Apis mellifera)',
        'view': 'lateral view',
        'features': 'fuzzy body, transparent wings, pollen baskets on legs, distinctive stripes, detailed wing venation',
        'color': 'golden brown and black'
    },
    '귀뚜라미': {
        'name': 'Field cricket (Gryllus campestris)',
        'view': 'lateral view',
        'features': 'long antennae, jumping hind legs, wing covers with sound-producing structures, detailed body segments',
        'color': 'dark brown'
    },
    '방아깨비': {
        'name': 'Rice grasshopper (Oxya japonica)',
        'view': 'lateral view',
        'features': 'large hind legs, detailed wing texture, segmented antennae, compound eyes, body segments',
        'color': 'green to brown'
    },
    '하늘소': {
        'name': 'Asian long-horned beetle (Anoplophora glabripennis)',
        'view': 'dorsal view',
        'features': 'extremely long antennae, distinctive wing covers, strong legs, detailed body segments, compound eyes',
        'color': 'black with white spots'
    }
}

def main():
    # Generate non-insect images
    print("\nGenerating non-insect images...")
    for name, details in not_insects.items():
        print(f"\nGenerating image for {name}...")
        try:
            generate_image(name, details, is_insect=False)
        except Exception as e:
            print(f"Failed to generate {name}: {str(e)}")
    
    # Generate insect images
    print("\nGenerating insect images...")
    for name, details in insects.items():
        print(f"\nGenerating image for {name}...")
        try:
            generate_image(name, details, is_insect=True)
        except Exception as e:
            print(f"Failed to generate {name}: {str(e)}")

if __name__ == "__main__":
    main() 